### ISSUE TITLE
memory free side-effect on MemRef value not supported!

### ISSUE THREAD

**Issue by nhasabni** [plaidml/tpp-mlir#769]
Tpp-opt is introducing `memref.dealloc` calls that seems to be leading to Memref value not supported error. This is for an MLP model from PyTorch.

**tpp-mlir Commit**
`5bba2907effe2d34fbb6c6af50f5eff8cb7e1ce9`

**Steps to reproduce**

```
$ tpp-opt -default-tpp-passes linear.mlir.txt > linear.tppopt.mlir.txt 
$ tpp-run linear.tppopt.mlir.txt -entry-point-result=void -e=forward
```
[linear.mlir.txt](https://github.com/plaidml/tpp-mlir/files/13273059/linear.mlir.txt)
[linear.tppopt.mlir.txt](https://github.com/plaidml/tpp-mlir/files/13273061/linear.tppopt.mlir.txt)

**Comment by rengolin** [comment]
Hm, the `ownership-based-buffer-deallocation` pass is adding `bufferization.dealloc` (and later `memref.alloc` and `memref.dealloc` pairs) inside the `scf.forall` loop, this is bad. And this is what `tpp-run` is complaining about:

```
$ ./bin/tpp-run -entry-point-result=void -e=forward linear.tppopt.mlir.txt 
loc("linear.tppopt.mlir.txt":64:7): error: memory free side-effect on MemRef value not supported!
ERROR: Failed to lower IR to LLVM dialect
```

Interestingly enough, if you just pass the original file to `tpp-run` (which also runs the default TPP passes, the crash does not happen.

@adam-smnk, there's some difference between the two "default" pipelines.

**Comment by nhasabni** [comment]
Seems that this is not a real issue. If I run mlir directly with `tpp-run`, it works fine.

**Comment by nhasabni** [comment]
Oops, I saw the discussion on the issue. So keeping it open for now. Sorry closing it early. :)

**Comment by adam-smnk** [comment]
The reproducer can be reduced to:
`tpp-opt -bufferize -bufferize linear.mlir.txt`
No crashes with:
`tpp-opt -bufferize -bufferize=dealloc=false linear.mlir.txt`

The first run of the `bufferize` pass (or in the original case `-default-tpp-passes`) inserts deallocations.
Running `buildBufferDeallocationPipeline` in presence of `memref.dealloc`s (second `-bufferize` or `tpp-run` on preprocessed IR) leads to the reported error.

**Comment by rengolin** [comment]
> Running buildBufferDeallocationPipeline in presence of memref.deallocs (second -bufferize or tpp-run on preprocessed IR) leads to the reported error.

Yup. I ran the debug scripts and bisected the pass. Essentially, the gist is that the `pack` is broken. It doesn't quite pack in reasonable shapes and when the `tile-and-fuse` passes, it moves the `tensor.empty` inside the loop, which creates the `memref.alloc` for the two sub-buffers.

```mlir
      // Transpose
      %alloc_10 = memref.alloc() {alignment = 64 : i64} : memref<128x32xf32>
      linalg.generic {indexing_maps = [#map, #map1], iterator_types = ["parallel", "parallel"]} ins(%subview_9 : memref<32x128xf32, strided<[128, 1], offset: ?>>) outs(%alloc_10 : memref<128x32xf32>) {
      ^bb0(%in: f32, %out: f32):
        linalg.yield %in : f32
      }
```

and 

```mlir
      // XSMM Zero local tiles instead of subview
      %alloc_11 = memref.alloc() {alignment = 64 : i64} : memref<2x32xf32>
      %intptr_12 = memref.extract_aligned_pointer_as_index %alloc_11 : memref<2x32xf32> -> index
      %14 = arith.index_cast %intptr_12 : index to i64
      %15 = llvm.inttoptr %14 : i64 to !llvm.ptr<f32>
      func.call @xsmm_unary_scalar_invoke(%c1_i64, %0, %cst, %15, %c0) : (i64, i64, f32, !llvm.ptr<f32>, index) -> ()
```

Another way to reproduce the error is to pipe the result back into `tpp-opt`:

```
$ ../bin/tpp-opt -default-tpp-passes ~/Downloads/linear.mlir.txt | ../bin/tpp-opt --default-tpp-passes
<stdin>:64:7: error: memory free side-effect on MemRef value not supported!
      memref.dealloc %alloc_10 : memref<128x32xf32>
      ^
<stdin>:64:7: note: see current operation: "memref.dealloc"(%48) : (memref<128x32xf32>) -> ()
```

Running a dump and a split shows me that, while the first pass runs `tile-and-fuse` once, the second pass runs 7 times and the pipeline is twice as long (216 steps instead of 99), so something is fishy on the default pass pipeline.

**Comment by rengolin** [comment]
Looking at the second pass dump, each pass runs multiple times before going to the next but actually don't change anything. Normally multiple iterations mean some dual fixed point flipping, but in this case there is no change at all.

The only thing that keeps flipping back and forth (across multiple and unrelated passes) is the map representation (from `#map` to inline back to `#map`).

The main difference between the previous IR and the one at the ownership deallocation is a bunch of `arith.constant` true/false inside the inner loop, which makes no sense at all:
```mlir
    ...
    scf.parallel (%arg5) = (%c0) to (%c256) step (%c32) {
      //  ¯\_(ツ)_/¯
      %true = arith.constant true
      %true_9 = arith.constant true
      %false = arith.constant false
      %false_10 = arith.constant false
      %false_11 = arith.constant false
      %false_12 = arith.constant false
      %false_13 = arith.constant false
      %subview = memref.subview %arg1[%arg5] [32] [1] : memref<256xf32> to memref<32xf32, strided<[1], offset: ?>>
      %false_14 = arith.constant false
      %subview_15 = memref.subview %arg0[%arg5, 0] [32, 128] [1, 1] : memref<256x128xf32> to memref<32x128xf32, strided<[128, 1], offset: ?>>
      %alloc_16 = memref.alloc() {alignment = 64 : i64} : memref<128x32xf32>
      linalg.generic {indexing_maps = [#map, #map1], iterator_types = ["parallel", "parallel"]} ins(%subview_15 : memref<32x128xf32, strided<[128, 1], offset: ?>>) outs(%alloc_16 : memref<128x32xf32>) {
      ^bb0(%in: f32, %out: f32):
        linalg.yield %in : f32
      }
      ...
```

This is also literally the _only_ difference between the input and the intermediate state at that pass, so it seems to be definitely a bug in the pass itself, not our code or the IR.

**Comment by rengolin** [comment]
In discussion with @nhasabni yesterday, the pass default pass pipeline seems ok, the problem is the ownership deallocation pass, as it cannot cope with its own output. The error seems to be that the pass cannot accept input with automatic deallocation and fails hard in the face of it. Our proposal is to change upstream from an error to just return pass failure and not modify the IR.

**Comment by matthias-springer** [comment]
Are you actually running the buffer deallocation pass multiple times or do you have pre-existing deallocs that were inserted by some other pass? It it's the former, I'd recommend to run the buffer deallocation pass a single time at the end. (It will also be more efficient.)

**Comment by rengolin** [comment]
> Are you actually running the buffer deallocation pass multiple times or do you have pre-existing deallocs that were inserted by some other pass?

Neither. We have two tools: `tpp-opt` and `tpp-run`. The former just run the passes, the latter run the "default pipeline", lowers to LLVM and execute.

In itself, this isn't the problem, we can very easily not pipe one into another. The original report was based on that mistaken usage, not a real use case. So this issue isn't "important" as is.

However, we have long-term plans to start embedding our experiments into other compilers. Our experience with IREE has shown us that it's impractical to try to control other tools' pipelines, and there's resistance to a truly flexible pipeline builder, so we settled on a strategy that passes MLIR as "messages", essentially piping the code through multiple tools, like old school compilers.

But in doing so, and assuming the other tools accept this model, it will still be hard to compose passes if they're not reentrant. For example, the producer tool runs the ownership pass and the consumer also (after some allocations of its own). Most other passes work fine, this is the first one we noticed doesn't. 

The main point in starting this discussion is to understand if this is a property we want passes to have (since most of them do) and if so, we should design all passes to be and have tests to make sure they are. This includes one-shot bufferization (which does mostly work) but does not include partial bufferization, which leaves the IR in a temporary perhaps unstable state. 

In light of more and more tools using core MLIR passes, and in our push to upstream as many as we can, we need to make sure composition of passes as well as integration of tools follow reasonable assumptions, or we'll end up with a broken upstream state and high downstream maintenance cost.

> I'd recommend to run the buffer deallocation pass a single time at the end. (It will also be more efficient.)

We do that in our default pipeline. 

The problem here is if you do: `tpp-opt -default-tpp-passes | tpp-run`, which then passes the default pipeline twice. It's not a "reasonable case", so I want to make sure that this isn't the source of the discussion, just the assumption if passes need to be reentrant for composition and stability promises.
